sdcm/utils/gce_builder.py

```diff
@@ -0,0 +1,239 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2023 ScyllaDB
+import os
+import sys
+import logging
+from typing import Any
+from functools import cached_property
+
+import requests
+from google.oauth2 import service_account
+from google.cloud import compute_v1
+from google.api_core.extended_operation import ExtendedOperation
+import google.api_core.exceptions
+
+from sdcm.utils.gce_region import GceRegion
+from sdcm.utils.gce_utils import SUPPORTED_PROJECTS, SUPPORTED_REGIONS
+from sdcm.sct_runner import GceSctRunner
+from sdcm.keystore import KeyStore
+
+LOGGER = logging.getLogger(__name__)
+
+JENKINS_CONFIG_FORMAT = """
+// just modify this config other code just logic
+config = [
+    name: "{name}",
+    region: "{region_name}",
+    projectId: "{project_id}",
+    credentialsId: "{project_id}",
+    instanceCapStr: "{max_instances}",
+    labels: "{jenkins_labels}",
+]
+
+import com.google.jenkins.plugins.computeengine.ComputeEngineCloud
+import com.google.jenkins.plugins.computeengine.InstanceConfiguration
+import com.google.jenkins.plugins.computeengine.SshConfiguration
+import hudson.plugins.sshslaves.SSHConnector
+import hudson.plugins.sshslaves.verifiers.NonVerifyingKeyVerificationStrategy
+import hudson.model.*
+import jenkins.model.Jenkins
+
+ComputeEngineCloud gcpCloud = new ComputeEngineCloud(config.name, config.projectId, config.credentialsId, config.instanceCapStr)
+gcpCloud.setNoDelayProvisioning(true)
+
+InstanceConfiguration instanceConfiguration = new InstanceConfiguration()
+instanceConfiguration.setNumExecutorsStr("4")
+instanceConfiguration.setOneShot(false)
+instanceConfiguration.setNamePrefix("builders")
+instanceConfiguration.setDescription("SCT Jenkins builders")
+instanceConfiguration.setRegion("https://www.googleapis.com/compute/v1/projects/{project_id}/regions/{region_name}")
+instanceConfiguration.setZone("https://www.googleapis.com/compute/v1/projects/{project_id}/zones/{region_name}-b")
+instanceConfiguration.setLabelString(config.labels)
+instanceConfiguration.setMode(Node.Mode.NORMAL)
+instanceConfiguration.setTemplate("https://www.googleapis.com/compute/v1/projects/{project_id}/global/instanceTemplates/{name}")
+instanceConfiguration.setRetentionTimeMinutesStr("6")
+instanceConfiguration.setLaunchTimeoutSecondsStr("300")
+instanceConfiguration.setRunAsUser("jenkins")
+instanceConfiguration.setSshConfiguration(SshConfiguration.builder().customPrivateKeyCredentialsId("6e17c350-0a47-4a59-b0b9-7e64f1c35a65").build())
```

Comment:
Shouldn't the creds ID be in some config? Especially when you say above that here we have `just logic`?

Reply:
it's only cause the key we need doesn't have meaningful name in Jenkins
we have keys similar to this one scattered in our pipelines as well (see `b8a774da-0e46-4c91-9f74-09caebaea261`)

it doesn't belongs into sct configuration, the only place I might thing of is our s3 bucket.
anyhow this would be tied to what's configured in our Jenkins server, regardless of where we'll put it.

Reply:
o.k.